main.go

```diff
@@ -72,13 +78,63 @@ func main() {
 	flag.Parse()
 
 	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+	id, err := os.Hostname()
+	if err != nil {
+		setupLog.Error(err, "error getting hostname", "controller", "Pod")
+		os.Exit(1)
+	}
+	kubeconfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+	namespace, _, err := kubeconfig.Namespace()
+	if err != nil {
+		setupLog.Error(err, "could not obtain a namespace to use for the leader election lock", "controller", id)
+		os.Exit(1)
+	}
+	config := ctrl.GetConfigOrDie()
+	leClientSet := kubernetes.NewForConfigOrDie(rest.AddUserAgent(config, "leader-election"))
 
+	// Lock required for leader election
+	rl, err := resourcelock.New(resourcelock.ConfigMapsLeasesResourceLock, namespace, "multiarch-operator-lock",
+		leClientSet.CoreV1(), leClientSet.CoordinationV1(), resourcelock.ResourceLockConfig{
+			Identity: id,
+		})
+	if err != nil {
+		setupLog.Error(err, "error creating leader election resource lock", "controller", id)
+		os.Exit(1)
+	}
+	leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
+		Lock:          rl,
+		LeaseDuration: 15 * time.Second,
+		RenewDeadline: 10 * time.Second,
+		RetryPeriod:   2 * time.Second,
+		Callbacks: leaderelection.LeaderCallbacks{
+			OnStartedLeading: func(c context.Context) {
+				run(metricsAddr, probeAddr)
+			},
+			OnStoppedLeading: func() {
+				setupLog.Info("no longer the leader, staying inactive.", "controller", id)
+			},
+			OnNewLeader: func(current_id string) {
+				if current_id == id {
+					setupLog.Info("i am the new leader!", "controller", id)
+					return
+				}
+				setupLog.Info("new leader is", "controller", current_id)
+			},
+		},
+	})
+
+}
+
+func run(metricsAddr string, probeAddr string) {
 	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
 		Scheme:                 scheme,
 		MetricsBindAddress:     metricsAddr,
 		Port:                   9443,
 		HealthProbeBindAddress: probeAddr,
-		LeaderElection:         enableLeaderElection,
+		LeaderElection:         true,
```

Comment:
do we need this?

Reply:
yes, we do. We need to be able disabling leader election in some cases at runtime: for example, the mutating webhook controller does not need leader election and a primary-backup setup. Rather, we want it to be replicated as a non-singleton, stateless service.

Reply:
that's true in this case because the webhook and the controller are one. In the real implementation they will run independently so we should be fine?